### ISSUE TITLE
FastChainBodySyncer crashes(?) and sync stalls

### ISSUE THREAD

**Issue by lithp** [ethereum/trinity#119]
* py-evm Version: c4e38746

### What is wrong?

```

DEBUG  01-07 13:57:40   FastChainBodySyncer  Duplicate headers during fast sync (<BlockHeader #214104 d837fccd>, <BlockHeader #214105 d1aaf1c3>, <BlockHeader #214106 cf8
[this line is insanely long, it includes 10929 instances of "BlockHeader"]

[40 minutes where the only lines look like this]
DEBUG  01-07 13:57:43     DiscoveryProtocol  Ignoring find_node request from unknown node <Node(0x77c7@172.104.162.211)>
DEBUG  01-07 13:57:44     DiscoveryProtocol  Ignoring find_node request from unknown node <Node(0x9043@142.93.125.215)>
DEBUG  01-07 13:57:45     DiscoveryProtocol  Ignoring find_node request from unknown node <Node(0x80a3@54.37.85.202)>
[the 40 minutes end]

    DEBUG  01-07 14:36:20   FastChainBodySyncer  <coroutine object FastChainBodySyncer._launch_prerequisite_tasks at 0x7f18b84c1a48> finished while <trinity.sync.full.chain.
FastChainBodySyncer object at 0x7f18b84c2cf8> is still running, terminating as well
ESC[1mESC[33m WARNING  01-07 14:36:20   FastChainBodySyncer  Task <coroutine object FastChainBodySyncer._launch_prerequisite_tasks at 0x7f18b82686c8> finished unexpectedly:
 Cannot prepare task <CompletionForBlockPersistPrereqs(<BlockHeader #136877 f3590012>, done=set(), remaining={<BlockPersistPrereqs.StoreBlockBodies: 1>, <BlockPersistPrereq
s.StoreReceipts: 2>})> with id b'\xf3Y\x00\x12I$\x15\'*=\rr@\xcbBe\x93\x1fj^\x10b\x07\x1e:\x85"\x89<\x04\x0fX' and dependency b'\x04I\xdf[>\xf1\xda\\iN\x82f\x94\x13\x80\xf1
C\x8a\\\xf0h\x0fn\xd9\xbc\xcd/\xd0\xf1nn\x0c' before preparing its dependencyESC[0m
   DEBUG  01-07 14:36:20   FastChainBodySyncer  Task failure traceback
Traceback (most recent call last):
  File "/home/brian/ef/trinity/trinity/sync/full/chain.py", line 446, in _launch_prerequisite_tasks
    self._block_persist_tracker.register_tasks(headers)
  File "/home/brian/ef/trinity/trinity/_utils/datastructures.py", line 534, in register_tasks
    duplicates,
trinity._utils.datastructures.DuplicateTasks: Cannot re-register tasks: (<BlockHeader #214104 d837fccd>, <BlockHeader #214105 d1aaf1c3>, <BlockHeader #214106 cf8126e3>, <Bl

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/brian/ef/trinity/p2p/service.py", line 141, in _run_task_wrapper
    await awaitable
  File "/home/brian/ef/trinity/p2p/service.py", line 160, in _run_daemon_task_wrapper
    await awaitable
  File "/home/brian/ef/trinity/trinity/sync/full/chain.py", line 452, in _launch_prerequisite_tasks
    self._block_persist_tracker.register_tasks(non_duplicates)
  File "/home/brian/ef/trinity/trinity/_utils/datastructures.py", line 547, in register_tasks
    f"Cannot prepare task {prereq_tracker!r} with id {task_id} and "
trinity._utils.datastructures.MissingDependency: Cannot prepare task <CompletionForBlockPersistPrereqs(<BlockHeader #136877 f3590012>, done=set(), remaining={<BlockPersistPrereqs.StoreBlockBodies: 1>, <BlockPersistPrereqs.StoreReceipts: 2>})> with id b'\xf3Y\x00\x12I$\x15\'*=\rr@\xcbBe\x93\x1fj^\x10b\x07\x1e:\x85"\x89<\x04\x0fX' and dependency b'\x04I\xdf[>\xf1\xda\\iN\x82f\x94\x13\x80\xf1C\x8a\\\xf0h\x0fn\xd9\xbc\xcd/\xd0\xf1nn\x0c' before preparing its dependency
Traceback (most recent call last):
  File "/home/brian/ef/trinity/trinity/sync/full/chain.py", line 446, in _launch_prerequisite_tasks
    self._block_persist_tracker.register_tasks(headers)
  File "/home/brian/ef/trinity/trinity/_utils/datastructures.py", line 534, in register_tasks
    duplicates,
trinity._utils.datastructures.DuplicateTasks: Cannot re-register tasks: (<BlockHeader #214104 d837fccd>, <BlockHeader #214105 d1aaf1c3>, <BlockHeader #214106 cf8126e3>, <Bl

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/brian/ef/trinity/p2p/service.py", line 141, in _run_task_wrapper
    await awaitable
  File "/home/brian/ef/trinity/p2p/service.py", line 160, in _run_daemon_task_wrapper
    await awaitable
  File "/home/brian/ef/trinity/trinity/sync/full/chain.py", line 452, in _launch_prerequisite_tasks
    self._block_persist_tracker.register_tasks(non_duplicates)
  File "/home/brian/ef/trinity/trinity/_utils/datastructures.py", line 547, in register_tasks
    f"Cannot prepare task {prereq_tracker!r} with id {task_id} and "
trinity._utils.datastructures.MissingDependency: Cannot prepare task <CompletionForBlockPersistPrereqs(<BlockHeader #136877 f3590012>, done=set(), remaining={<BlockPersistPrereqs.StoreBlockBodies: 1>, <BlockPersistPrereqs.StoreReceipts: 2>})> with id b'\xf3Y\x00\x12I$\x15\'*=\rr@\xcbBe\x93\x1fj^\x10b\x07\x1e:\x85"\x89<\x04\x0fX' and dependency b'\x04I\xdf[>\xf1\xda\\iN\x82f\x94\x13\x80\xf1C\x8a\\\xf0h\x0fn\xd9\xbc\xcd/\xd0\xf1nn\x0c' before preparing its dependency
```

After this log output nothing really happens. Peer discovery continues, requests from remote nodes are answered. A few hours later `HeaderMeatSyncer` throws an exception complaining about stale blocks.

Some second-order problems:
- [x] If FastChainBodySyncer crashes (and probably any of a set of critical services) the entire client should shut down. Someone who really cares can detect and react to a lack of progress but even trivial monitoring setups will notice a process ending.
- [x] The BlockHeader lines really clutter up the log, it's nearly impossible to scroll past them using `less` unless you disable line wrapping. These lines in particular should be clipped, and maybe the logger should clip all lines. 10,000 chars should be enough for anybody ;)

**Comment by veox** [comment]
Confirming having observed the first part (but only that) the last week: duplicate headers log message (with very long output), and sync stall coinciding with that.

**Comment by veox** [comment]
A bit of debugging here, as follows:

``` python
        try:
            self._block_persist_tracker.register_tasks(headers)
        except DuplicateTasks as exc:
            # Likely scenario: switched which peer downloads headers, and the new peer isn't                       
            # aware of some of the in-progress headers                                                             
            self.logger.debug(
                "Duplicate headers during fast sync (%s total), skipping...",
                len(exc.duplicates)
            )
            non_duplicates = tuple(header for header in headers if header not in exc.duplicates)
            self.logger.debug(
                "Non-duplicate headers selected (%s total), registering tasks...",
                len(non_duplicates)
            )
            self._block_persist_tracker.register_tasks(non_duplicates)
            self.logger.debug("Done registering non-duplicate header tasks.")
```

... picks out the following log lines:

```
   DEBUG  01-14 17:15:16   FastChainBodySyncer  Duplicate headers during fast sync (3670 total), skipping...

<Discovery keeps running for 5 minutes...>

   DEBUG  01-14 17:19:14   FastChainBodySyncer  Non-duplicate headers selected (66119 total), registering tasks...
   DEBUG  01-14 17:19:15   FastChainBodySyncer  <coroutine object FastChainBodySyncer._launch_prerequisite_tasks at 0x7f3661263f48> finished while <trinity.sync.full.chain.FastChainBodySyncer object at 0x7f36609fd710> is still running, te\
rminating as well
 WARNING  01-14 17:19:15   FastChainBodySyncer  Task <coroutine object FastChainBodySyncer._launch_prerequisite_tasks at 0x7f36609f3bc8> finished unexpectedly: Cannot prepare task <CompletionForBlockPersistPrereqs(<BlockHeader \
#2773293 aed309cb>, done=set(), remaining={<BlockPersistPrereqs.StoreReceipts: 2>, <BlockPersistPrereqs.StoreBlockBodies: 1>})> with id b'\xae\xd3\t\xcbb#\xc3\xe8vpL\x08\x18<*\xfa\xf1\x95\xb0\x80\xec|\x96`\xbf\xca\xadl\x0c\x10\xae\xef' an\
d dependency b'W\xe3\xe7{\x16\x027\xfc\xf2\x1f6KU\xdd\xa0\x83\xee\x9f\xd4\xf9:\xd2\x82O\x1c\xac\xe8q\xdf\xff\x0e\xdb' before preparing its dependency
   DEBUG  01-14 17:19:15   FastChainBodySyncer  Task failure traceback
Traceback (most recent call last):
  File "/home/veox/src/trinity/trinity/sync/full/chain.py", line 446, in _launch_prerequisite_tasks
    self._block_persist_tracker.register_tasks(headers)
  File "/home/veox/src/trinity/trinity/_utils/datastructures.py", line 534, in register_tasks
    duplicates,
trinity._utils.datastructures.DuplicateTasks: Cannot re-register tasks: (<another awfully-long line listing many elements of form <BlockHeader #2805685 8ebb8ae8>>) for completion

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/veox/src/trinity/p2p/service.py", line 141, in _run_task_wrapper
    await awaitable
  File "/home/veox/src/trinity/p2p/service.py", line 160, in _run_daemon_task_wrapper
    await awaitable
  File "/home/veox/src/trinity/trinity/sync/full/chain.py", line 459, in _launch_prerequisite_tasks
    self._block_persist_tracker.register_tasks(non_duplicates)
  File "/home/veox/src/trinity/trinity/_utils/datastructures.py", line 547, in register_tasks
    f"Cannot prepare task {prereq_tracker!r} with id {task_id} and "
trinity._utils.datastructures.MissingDependency: Cannot prepare task <CompletionForBlockPersistPrereqs(<BlockHeader #2773293 aed309cb>, done=set(), remaining={<BlockPersistPrereqs.StoreReceipts: 2>, <BlockPersistPrereqs.StoreBlockBodies: \
1>})> with id b'\xae\xd3\t\xcbb#\xc3\xe8vpL\x08\x18<*\xfa\xf1\x95\xb0\x80\xec|\x96`\xbf\xca\xadl\x0c\x10\xae\xef' and dependency b'W\xe3\xe7{\x16\x027\xfc\xf2\x1f6KU\xdd\xa0\x83\xee\x9f\xd4\xf9:\xd2\x82O\x1c\xac\xe8q\xdf\xff\x0e\xdb' befo\
re preparing its dependency
```

I stopped the process at this point to inspect; didn't check what happens afterwards (yet).

**Comment by veox** [comment]
It would seem that in `trinity/sync/full/chain.py`, non-duplicates are selected, and passed to `trinity/_utils/datastructures.py`.

That raises the exception saying there are duplicates...

-----

Those long lines I so frivolously shortened might actually be useful, in non-truncated form. Maybe they should be pushed to log level `DEBUG2`, or something...

**Comment by veox** [comment]
Cc: @carver (since you wrote much of `trinity/_utils/datastructures.py`):

It seems that `FastChainBodySyncer` can't recover from underlying exceptions if, upon switching peer being synced from, there are duplicates in header-syncing tasks.

There seems to be some sort of mix-up in re-adding tasks with missing (completed?..) dependencies. (Very much guessing here.)

**Comment by veox** [comment]
Looking again at the original error, and https://github.com/ethereum/trinity/blob/0dd33ee304630b93192861626ac5e9eca6fc4d92/trinity/sync/full/chain.py#L453 - there _is_ some handling of `MissingDependency`... Perhaps where it's not caught?..

**Comment by carver** [comment]
Thanks for mentioning me. I'll definitely take a look today.

This code went through a big re-work recently and still has some significant work to do, so no big surprise that there are bugs. Sorry that it's getting in your guys' way.

**Comment by carver** [comment]
> If FastChainBodySyncer crashes (and probably any of a set of critical services) the entire client should shut down. Someone who really cares can detect and react to a lack of progress but even trivial monitoring setups will notice a process ending.

Hm, this is a bit of a surprise by just reading the code. It looks like the whole chain of services up to the `FullNodeSyncer` is using a `run_daemon_task()` or a simple `await service.run()`. I would expect both of those to trigger fail outs. In fact, you can see the first one here:

```
    DEBUG  01-07 14:36:20   FastChainBodySyncer  <coroutine object FastChainBodySyncer._launch_prerequisite_tasks at 0x7f18b84c1a48> finished while <trinity.sync.full.chain.FastChainBodySyncer object at 0x7f18b84c2cf8> is still running, terminating as well
```

So I'll probably have to play a bit to figure it out.

**Comment by veox** [comment]
> If FastChainBodySyncer crashes (and probably any of a set of critical services) the entire client should shut down.

Did this get fixed by PR #166?.. From skinning its diff it seems it was supposed to. Rebased work branch to check...

**Comment by carver** [comment]
> > If FastChainBodySyncer crashes (and probably any of a set of critical services) the entire client should shut down.
> 
> Did this get fixed by PR #166?.. From skimming its diff it seems it was supposed to. Rebased work branch to check...

Yes (at least that was the intention, and it works in simple local tests). I just checked that one off.

@lithp if you can give a :+1: to  #165 then I think we can get this whole issue closed.

**Comment by lithp** [comment]
@carver sorry for the delay on #165, thanks for the ping! I'll review it tomorrow, it looks good to me but I want to understand the code a little better before saying :shipit:
